### PR TITLE
Fixed #19361 -- link to object added to after-save message

### PR DESCRIPTION
**Pull request by bak1an**
rebased old & dusty patch for https://code.djangoproject.com/ticket/19361

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
I think it might be better to use [`format_html()`](https://docs.djangoproject.com/en/1.8/ref/utils/#django.utils.html.format_html) to build the message rather than simply calling `mark_safe()` at the end.

**Comment by bak1an** [comment]
@timgraham Thanks for a fast review!

This is troubling me as well, but there is also a translation involved, like `msg = _('The %(name)s "%(obj)s" was changed successfully.') % msg_dict`, where `obj` key at `msg_dict` contains actual link snippet. 
Is there a way to mark a part of translated string to be safe (just a link) without marking the whole string? Splitting it into parts seems like a huge mess (requires updating all translations at least)

On the other hand - link is url encoded, link text is escaped and translations itself is unlikely to contain something dangerous, so whole string seems to be "safe". Might need to escape `name` key though, maybe with switch to `format_html()` there - https://github.com/django/django/pull/5244/files#diff-3c42de3e53aba87b32c494f995a728dfR1067 

What do you think?

**Comment by timgraham** [comment]
Translation isn't my expertise, but could https://docs.djangoproject.com/en/1.8/topics/i18n/translation/#other-uses-of-lazy-in-delayed-translations help?

File: django/contrib/admin/options.py [hunk 1]
```diff
@@ -1058,7 +1058,14 @@ def response_add(self, request, obj, post_url_continue=None):
         opts = obj._meta
         pk_value = obj._get_pk_val()
         preserved_filters = self.get_preserved_filters(request)
-        msg_dict = {'name': force_text(opts.verbose_name), 'obj': force_text(obj)}
+        obj_url = reverse('admin:%s_%s_change' %
```
Comment (timgraham) [inline django/contrib/admin/options.py]: Please use a style like this:
```python
reverse(
    'admin:%s_%s_change' % (opts.app_label, opts.model_name),
    args=(quote(pk_value),),
    current_app=self.admin_site.name,
)
```

File: django/contrib/admin/options.py [hunk 2]
```diff
@@ -1122,17 +1126,20 @@ def response_change(self, request, obj):
         pk_value = obj._get_pk_val()
         preserved_filters = self.get_preserved_filters(request)
 
-        msg_dict = {'name': force_text(opts.verbose_name), 'obj': force_text(obj)}
+        msg_dict = {
+            'name': force_text(opts.verbose_name),
+            'obj': force_text('<a href="{0}">{1}</a>').format(urlquote(request.path), escape(obj))
```
Comment (timgraham) [inline django/contrib/admin/options.py]: Please including a trailing comma in the last item of a dictionary so if more items are added we don't need to modify this line again.

File: tests/admin_views/tests.py [hunk 3]
```diff
@@ -3743,11 +3786,12 @@ def test_add_model_modeladmin_defer_qs(self):
                                     post_data, follow=True)
         self.assertEqual(response.status_code, 200)
         self.assertEqual(CoverLetter.objects.count(), 1)
-        # Message should contain non-ugly model verbose name
+        # Message should contain non-ugly model verbose name and link to edit form
```
Comment (timgraham) [inline tests/admin_views/tests.py]: Please add period at end of sentence. Comments should be wrapped at 79 characters too.

File: tests/admin_views/tests.py [hunk 4]
```diff
@@ -3850,8 +3897,8 @@ def test_edit_model_modeladmin_defer_qs(self):
         # instance representation is set by six.text_type()
         self.assertContains(
             response,
-            '<li class="success">The short message &quot;ShortMessage_Deferred_timestamp object&quot; was changed successfully.</li>',
-            html=True
+            '<li class="success">The short message "<a href="%s">ShortMessage_Deferred_timestamp object</a>" was changed successfully.</li>' %
```
Comment (timgraham) [inline tests/admin_views/tests.py]: please limit line lengths to 119 characters like this:
```
'<li class="success">The short message "<a href="%s">ShortMessage_Deferred_timestamp object</a>" '
'was changed ....'
```

File: tests/admin_views/tests.py [hunk 5]
```diff
@@ -789,6 +789,49 @@ def test_allows_attributeerror_to_bubble_up(self):
         with self.assertRaises(AttributeError):
             self.client.get(reverse('admin:admin_views_simple_changelist'))
 
+    def test_after_add_message_contains_change_link(self):
```
Comment (timgraham) [inline tests/admin_views/tests.py]: The new test seems a bit redundant with the other test changes. Does it test anything different?

**Comment by timgraham** [comment]
As far as documentation goes, at least a mention in the 1.9 release notes is appropriate. You could check `docs/ref/contrib/admin` and ensure nothing needs to be updated there?

File: django/contrib/admin/options.py [hunk 6]
```diff
@@ -1058,7 +1058,14 @@ def response_add(self, request, obj, post_url_continue=None):
...
         pk_value = obj._get_pk_val()
         preserved_filters = self.get_preserved_filters(request)
-        msg_dict = {'name': force_text(opts.verbose_name), 'obj': force_text(obj)}
+        obj_url = reverse('admin:%s_%s_change' %
+                          (opts.app_label, opts.model_name),
+                          args=(quote(pk_value),),
+                          current_app=self.admin_site.name)
+        msg_dict = {
+            'name': force_text(opts.verbose_name),
+            'obj': force_text('<a href="{0}">{1}</a>').format(urlquote(obj_url), escape(obj))
```
Comment (kezabelle) [inline django/contrib/admin/options.py]: Won't replacing `obj` in the message mean that when *Save and continue editing* is chosen the message will say something like `"saved <a href="...">obj</a>, you may continue editing it below"` where the `<a href>` is a link to the same page you're already on?

Comment (bak1an) [inline django/contrib/admin/options.py]: @kezabelle 
Indeed, when "Save and continue" is used to save object, link in post save message will be pointing to the very same page user is already on.

I think that it is better to leave it this way to not introduce any mess into "Post save message contains link to object's edit form" assertion. 
Besides that, this links does not do any harm (changed `color` and `cursor` css properties for object name does not look so scary) and even can be useful sometimes, like for copying the link on mobile devices, where address bar can often be hidden.

**Comment by bak1an** [comment]
>Translation isn't my expertise, but could https://docs.djangoproject.com/en/1.8/topics/i18n/translation/#other-uses-of-lazy-in-delayed-translations help?

Does not look so. Eventually this will resolve to the very same string marked as safe, so at rendering point (https://github.com/django/django/blob/master/django/contrib/admin/templates/admin/base.html#L65) it will be the same as it is now.

**Comment by bak1an** [comment]
>As far as documentation goes, at least a mention in the 1.9 release notes is appropriate. You could check docs/ref/contrib/admin and ensure nothing needs to be updated there?

Changelog entry added.

It does not look like `docs/ref/contrib/admin` needs updating. There are no screenshots with post-save message visible and post-save messages are not mentioned. It also does not look like this particular feature is big/significant enough to get its own documentation.
You see a link - you can click it.

**Comment by bak1an** [comment]
Hope this makes sense, switching ticket back to 'Patches needing review'

**Comment by bak1an** [comment]
Rebased again since there were conflicts with current `master`.

**Comment by timgraham** [comment]
@claudep, could you check if we should be handling the string translations some other way or if the suggested use of `mark_safe()` is okay? Thanks!

**Comment by claudep** [comment]
I tried the format_html way, see https://gist.github.com/claudep/aa1edac079866a482510. Indeed, the main drawback is changed strings for translators, but that happens.

**Comment by bak1an** [comment]
@timgraham 
@claudep 

My guess is that this won't be landing in 1.9 since it's freezed already so I will update release notes to 1.10 here.

I can also update the patch to use `format_html` but I'm going to need some guidance on how to deal with translations.

Should I apply all language teams at transifex to update strings everywhere or I can update .po files with this pull request? It is not really a 'translation' it is just a formatting change from `%(name)s "%(obj)s"` to `{name} "{obj}"` so it does not require me knowing all those languages.

Thanks.

**Comment by claudep** [comment]
No, translation updates are automatically picked up when translation templates are updated (pre-release time). No need to handle them in the patch.

**Comment by bak1an** [comment]
@timgraham 
It uses `format_html` now (thanks @claudep !). And release notes were updated for 1.10 instead of 1.9.

**Comment by timgraham** [comment]
Perhaps we should include the link only if the user has "change" permission for the object. Otherwise, if the user has only add permission, the link will lead to a page that returns a 403.

**Comment by bak1an** [comment]
Indeed.

Good catch, thanks!

On Wed, Sep 30, 2015 at 4:28 PM, Tim Graham <notifications@github.com>
wrote:

> Perhaps we should include the link only if the user has "change"
> permission for the object. Otherwise, if the user has only add permission,
> the link will lead to a page that returns a 403.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/django/django/pull/5244#issuecomment-144396058>.
>



-- 
Anton

**Comment by bak1an** [comment]
@timgraham done

**Comment by timgraham** [comment]
merged in 80bcbecd4a394f67c1679b077ba79a797c4a57d7, thanks!

**Comment by bak1an** [comment]
@timgraham  awesome! thanks!

**Pull request closed without merging**